tests/unit/lms/services/canvas_api/__init___test.py

```diff
@@ -0,0 +1,61 @@
+from unittest.mock import sentinel
+
+import pytest
+
+from lms.services.canvas_api import canvas_api_client_factory
+
+
+@pytest.mark.usefixtures("ai_getter")
+class TestCanvasAPIClientFactory:
```

Comment:
These tests are dumb. These really are the MD5 your code kind of tests. I don't know how to do this better in our style.

This would be easier if the attributes were public on the objects, then we could just instantiate one and interrogate that all the attributes. We could technically do that now with something like:

```python
assert result == Any.object(CanvasAPIClient).with_attrs({
    '_api': Any.object(AuthenticatedClient).with_attrs({
        ... 
    })

But it feels a bit dirty, and we don't need these attributes to be public for any other reason.
})
```

Reply:
I don't think this is a problem with our test style. I think the problem is that `__init__.py` doesn't actually contain anything very interesting to unit test. It's integration code, and that kind of code just isn't very interesting to test. You want your tests to run the code to make sure it isn't completely broken. But beyond that there's not much of value to test. All the `assert_called_once_with()`'s are indeed kind of pointless.

I don't think there's much difference between with-mocks vs without-mocks here. A without-mocks version of this test doesn't look any more meaningful than the with-mocks version to me:

```python
assert isinstance(result, CanvasAPIClient)
assert isinstance(result._api, AuthenticatedClient)
```

As you say you'd have to either test the private `_api` attr or make `_api` public just for the test's sake, both of which feel icky. And even then, is this really any more meaningful than:

```python
CanvasAPIClient.assert_called_once_with(AuthenticatedClient.return_value)
assert result == CanvasAPIClient.return_value
```

Both seem equally meaningless to me. Either way you're just testing that it passed the `AuthenticatedClient` object to `CanvasAPIClient()` and returned the result. Both feel like MD5ing your code.

Here's a couple of options:

### With mocks

This is what I would do. I would spend too much time writing or thinking about it, because there's not much to test here. The test isn't actually testing much, it's just exercising the code so that we know it's not completely broken. I would keep the mocks for reasons explained in **Without mocks** below. The `BasicClient` assertion seems worth keeping since that's testing one thing the method actually does do (with `netloc`):

```python
from unittest.mock import sentinel

import pytest

from lms.services.canvas_api import canvas_api_client_factory


class TestCanvasAPIClientFactory:
    def test_it(self, ai_getter, pyramid_request, BasicClient, CanvasAPIClient):
        ai_getter.lms_url.return_value = "https://example.com/path"

        canvas_api = canvas_api_client_factory(sentinel.context, pyramid_request)

        BasicClient.assert_called_once_with("example.com")
        assert canvas_api == CanvasAPIClient.return_value


pytestmark = pytest.mark.usefixtures("ai_getter")


@pytest.fixture(autouse=True)
def BasicClient(patch):
    return patch("lms.services.canvas_api.BasicClient")


@pytest.fixture(autouse=True)
def TokenStore(patch):
    return patch("lms.services.canvas_api.TokenStore")


@pytest.fixture(autouse=True)
def AuthenticatedClient(patch):
    return patch("lms.services.canvas_api.AuthenticatedClient")


@pytest.fixture(autouse=True)
def CanvasAPIClient(patch):
    return patch("lms.services.canvas_api.CanvasAPIClient")
```

This test:

* Will fail if `canvas_api_client_factory` looks up the `ai_getter` service incorrectly
* Will fail if it doesn't do the `netloc` stuff correctly and pass it to `BasicClient`
* Will fail if it passes the wrong number of arguments, or an unknown keyword argument, to `BasicClient`, `TokenStore`, `AuthenticatedClient` or `CanvasAPIClient`
* Will fail if it doesn't return the `CanvasAPIClient`

It _won't_ fail if all the arguments it passes to everything are all completely wrong but still happen to have the right number of arguments and keyword-argument names for each method. But as you said, testing that really just feels like MD5ing your code and doesn't seem to add much value.

You _could_ add those asserts back in. I don't think they're doing much harm. I would just add them all into the one test, rather than a separate test for each assert. But as you say they're not adding much value

```python
    def test_it(
        self,
        ai_getter,
        pyramid_request,
        BasicClient,
        TokenStore,
        AuthenticatedClient,
        CanvasAPIClient,
    ):
        ai_getter.lms_url.return_value = "https://example.com/path"

        canvas_api = canvas_api_client_factory(sentinel.context, pyramid_request)

        BasicClient.assert_called_once_with("example.com")
        TokenStore.assert_called_once_with(
            consumer_key=pyramid_request.lti_user.oauth_consumer_key,
            user_id=pyramid_request.lti_user.user_id,
            db=pyramid_request.db,
        )
        AuthenticatedClient.assert_called_once_with(
            basic_client=BasicClient.return_value,
            token_store=TokenStore.return_value,
            client_id=ai_getter.developer_key(),
            client_secret=ai_getter.developer_secret(),
            redirect_uri=pyramid_request.route_url("canvas_oauth_callback"),
        )
        CanvasAPIClient.assert_called_once_with(AuthenticatedClient.return_value)
        assert canvas_api == CanvasAPIClient.return_value
```

### Without mocks

If I were writing this test and I disliked mocks I would probably do this:

```python
from unittest.mock import sentinel

import pytest

from lms.services.canvas_api import canvas_api_client_factory, CanvasAPIClient
from lms.services.canvas_api.public import CanvasAPIClient


class TestCanvasAPIClientFactory:
    def test_it(self, pyramid_request):
        canvas_api = canvas_api_client_factory(sentinel.context, pyramid_request)

        assert isinstance(canvas_api, CanvasAPIClient)


pytestmark = pytest.mark.usefixtures("ai_getter")
```

It's quite hard to compare with with-mocks and without-mocks versions given such a simple test, but:

* I don't think the without-mocks version has a good way to test that the factory does the right thing with `netloc` and passes it to `BasicClient`
* They both seem to run in the same amount of time
* Without-mocks is easier to write: you don't need the patch fixtures. (With-mocks is often easier to write because constructing real objects and getting them to behave as you want is often hard, whereas with mocks that's easy. But in this test, that doesn't apply.)
* The without-mocks version is less isolated. The test has more reasons to fail. If there are bugs in the `__init__()` methods of `BasicClient`, `TokenStore`, `AuthenticatedClient` or `CanvasAPIClient` or anything they call, they could cause this test to fail, which would just add noise and distraction to the test output, make debugging harder. We only want this test to fail if there's a bug in `canvas_api_client_factory()`. Bugs in the other classes should only cause the other classes' unit tests to fail.
* The without-mocks version messes up coverage reporting, making it look like those `__init__()` methods and anything they call have unit test coverage when they don't really. This harms coverage's ability to reveal bugs or test mistakes for us

### A more meaningful integration test?

You could try to write a test constructs a `CanvasAPIClient` without mocking its dependencies and then actually make one API call with that client and test (using a mocked out `requests` lib or whatever) that it actually made the expected HTTP call and returned the response.

I think you'd want to have all the details already covered with unit tests for the different components such that the integration test really is just finally testing that it all fits together just so we can feel safer about that, and no more than one or two integration tests are needed. And I think we'd want to exclude the integration test from counting towards test coverage.